File: backend/app/schemas/tutor.py
import uuid

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import UTCDateTime
from app.schemas.subject import TutorSubjectOut

# Lowercase letters/digits, hyphens allowed in the middle only, 3-64 chars total -
# used for the tutor's optional public-profile nickname (it-tutor.pro/tutors/<slug>).
SLUG_PATTERN = r"^[a-z0-9](?:[a-z0-9-]{1,62}[a-z0-9])?$"


class TutorScheduleSettings(BaseModel):
    slot_granularity_minutes: int
    break_between_lessons_minutes: int = Field(ge=0)
    min_lead_time_hours: int = Field(ge=0)
    cancel_min_hours_before: int = Field(ge=0)
    cancel_max_per_month: int = Field(ge=0)
    reschedule_min_hours_before: int = Field(ge=0)
    reschedule_max_per_month: int = Field(ge=0)


class TutorProfileUpdate(BaseModel):
    about: str | None = None
    is_hidden: bool | None = None
    slug: str | None = Field(default=None, pattern=SLUG_PATTERN)
    slot_granularity_minutes: int | None = None
    break_between_lessons_minutes: int | None = Field(default=None, ge=0)
    min_lead_time_hours: int | None = Field(default=None, ge=0)
    cancel_min_hours_before: int | None = Field(default=None, ge=0)
    cancel_max_per_month: int | None = Field(default=None, ge=0)
    reschedule_min_hours_before: int | None = Field(default=None, ge=0)
    reschedule_max_per_month: int | None = Field(default=None, ge=0)


class TutorProfileOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    photo_url: str | None
    about: str
    is_hidden: bool
    slug: str | None = None
    slot_granularity_minutes: int
    break_between_lessons_minutes: int
    min_lead_time_hours: int
    cancel_min_hours_before: int
    cancel_max_per_month: int
    reschedule_min_hours_before: int
    reschedule_max_per_month: int
    display_name: str | None = None
    is_active: bool | None = None
    # Populated only where the caller already has the linked User at hand (own
    # profile, admin) - see api/v1/tutors.py::_to_profile_out /
    # api/v1/admin.py::_to_profile_out. Lets the edit form prefill the split name
    # fields and email instead of only showing the denormalized display_name.
    first_name: str | None = None
    last_name: str | None = None
    patronymic: str | None = None
    email: str | None = None
    # Populated only by the admin tutor listing (see api/v1/admin.py::list_tutors) -
    # lets the admin bookings-tab filters narrow bookings down by the tutor's subjects
    # and directions without a separate lookup.
    subjects: list[TutorSubjectOut] = Field(default_factory=list)


class TutorCatalogItem(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    display_name: str
    # "Имя Отчество" only (no surname) - the format used on catalog cards.
    name_patronymic: str
    photo_url: str | None
    slug: str | None = None
    subjects: list[str] = Field(default_factory=list)
    # Cheapest individual lesson type, normalized to a per-hour rate
    # (price / duration_minutes * 60) so tutors with different lesson lengths are
    # comparable - see tutor_service.search_catalog.
    hourly_price: float | None = None
    avg_rating: float | None = None
    reviews_count: int = 0


class TutorCatalogPageOut(BaseModel):
    items: list[TutorCatalogItem]
    total: int
    page: int
    page_size: int


class TutorPublicProfile(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    display_name: str
    photo_url: str | None
    about: str
    subjects: list[TutorSubjectOut] = Field(default_factory=list)
    avg_rating: float | None = None
    reviews_count: int = 0


class UploadedImageOut(BaseModel):
    """A single uploaded image's URL, e.g. for insertion into the "about" rich text
    field - see api/v1/tutors.py::upload_about_image."""

    url: str


class TutorStudentOut(BaseModel):
    """Entry in the tutor's homework-assignment student picker (see
    booking_service.list_students_for_tutor), sorted by most recent lesson first."""

    id: uuid.UUID
    first_name: str
    last_name: str
    grade: int | None
    last_lesson_at: UTCDateTime | None

File: frontend/src/types/tutor.ts
import type { TutorSubject } from "@/types/subject";

export interface TutorExtraLink {
  label: string;
  url: string;
}

export interface TutorProfile {
  id: string;
  user_id: string;
  photo_url: string | null;
  about: string;
  is_hidden: boolean;
  slug: string | null;
  telegram_url: string | null;
  vk_url: string | null;
  youtube_url: string | null;
  video_url: string | null;
  extra_links: TutorExtraLink[];
  slot_granularity_minutes: number;
  break_between_lessons_minutes: number;
  min_lead_time_hours: number;
  cancel_min_hours_before: number;
  cancel_max_per_month: number;
  reschedule_min_hours_before: number;
  reschedule_max_per_month: number;
  allow_individual_bookings: boolean;
  allow_group_bookings: boolean;
  display_name: string | null;
  is_active: boolean | null;
  // Populated only for own profile / admin contexts - see backend
  // api/v1/tutors.py and api/v1/admin.py's _to_profile_out.
  first_name?: string | null;
  last_name?: string | null;
  patronymic?: string | null;
  email?: string | null;
  email_verified?: boolean | null;
  // Populated only by the admin tutor listing - see backend api/v1/admin.py::list_tutors.
  subjects?: TutorSubject[];
}

export interface TutorCatalogItem {
  id: string;
  user_id: string;
  display_name: string;
  // "Имя Отчество" only (no surname) - catalog card format.
  name_patronymic: string;
  photo_url: string | null;
  slug: string | null;
  subjects: string[];
  hourly_price: number | null;
  avg_rating: number | null;
  reviews_count: number;
  about_snippet: string | null;
  show_individual_booking: boolean;
  show_group_booking: boolean;
}

export interface TutorPublicProfile {
  id: string;
  user_id: string;
  display_name: string;
  photo_url: string | null;
  about: string;
  telegram_url: string | null;
  vk_url: string | null;
  youtube_url: string | null;
  video_url: string | null;
  // Player URL built by the backend from video_url (app/utils/video.py) - rendered
  // as an <iframe src> as-is, the frontend never parses the tutor's link itself.
  video_embed_url: string | null;
  extra_links: TutorExtraLink[];
  subjects: TutorSubject[];
  avg_rating: number | null;
  reviews_count: number;
  // Already combine the tutor's toggle with whether they have an active lesson type
  // of that format - see backend api/v1/tutors.py::get_public_profile.
  show_individual_booking: boolean;
  show_group_booking: boolean;
}

export interface TutorStudent {
  id: string;
  first_name: string;
  last_name: string;
  grade: number | null;
  last_lesson_at: string | null;
}

export interface StudentGroupMembership {
  group_id: string;
  group_name: string;
  status: string;
  joined_at: string;
  left_at: string | null;
}

export interface TutorStudentDetail {
  id: string;
  first_name: string;
  last_name: string;
  patronymic: string | null;
  grade: number | null;
  email: string | null;
  groups: StudentGroupMembership[];
}

export interface LessonType {
  id: string;
  tutor_id: string;
  name: string;
  format: "individual" | "group";
  duration_minutes: number;
  price: number;
  is_active: boolean;
}

export interface AvailabilityInterval {
  id: string;
  weekday: number;
  start_time: string;
  end_time: string;
}

export interface Slot {
  start_at: string;
  end_at: string;
  available: boolean;
}
